### ISSUE TITLE
Add awareness of safeAreaLayoutGuide and safeAreaInsets

### ISSUE THREAD

**Issue by RickyRomero** [lionheart/KeyboardAdjuster#5]
When compiled against the iOS 11 SDK, KeyboardAdjuster isn't aware of the new Safe Area APIs. In my app, this caused elements to underlap the home indicator on iPhone X.

In order to support it in my app, I had to make a couple of changes:

1. In my view controller, I needed to change the `keyboardAdjusterConstraint` so that it was based on `view.safeAreaLayoutGuide.bottomAnchor` instead of `view.bottomAnchor`.
2. In KeyboardAdjuster.swift, under the `case .visible` condition, I had to add the following lines to cancel out the safe area offset when the keyboard was open:
```swift
let frame = value.cgRectValue
let keyboardFrameInViewCoordinates = view.convert(frame, from: nil)
var additionalOffset:CGFloat = 0

if #available(iOS 11, *) {
  additionalOffset -= view.safeAreaInsets.bottom
}

constraint.constant = view.bounds.height - keyboardFrameInViewCoordinates.origin.y + additionalOffset
```

I'm a newbie at iOS development though, so I don't know how robust this solution is. Regardless, it works perfectly for my purpose.

**Comment by dlo** [comment]
@RickyRomero Thanks for the report. Your solution looks good but let me get the tests running again and see whether this is robust enough. Feel free to submit a PR if you'd like to contribute! :)

**Comment by dlo** [comment]
After double checking all the code, I've confirmed that KeyboardAdjuster doesn't make any assumptions about the safe area layout guide one way or the other. It's just a README issue, but I'll update that soon. Whether you want the pinned view to appear below or at the safeAreaLayoutGuide is up to you :). I'll make a note of it though.

As for cancelling out the offset, it doesn't look like that's a problem as of the latest iOS. It must have just been a bug in earlier versions of iOS 11. Let me know if you're still seeing an issue but after some debugging, the keyboard frame looks correct in view coordinates without taking the offset into account on the iPhone X.

**Comment by RickyRomero** [comment]
Hi Dan,

Re: canceling out the offset... Here's the issue I see when invoking the keyboard.

![gap](https://user-images.githubusercontent.com/233891/33446643-90fef70e-d5b5-11e7-864d-0969227c08f3.gif)

Notice the extra gap between the input field and the keyboard. Now, if I add that snippet to KeyboardAdjuster.swift, the gap disappears (and smoothly animates to the correct position).

![fix](https://user-images.githubusercontent.com/233891/33446650-98ddf1e6-d5b5-11e7-8c1e-c390498f01fc.gif)

If you need, I can try and provide a simplified test case for you.

Of course, my solution hardcodes an assumption about the safe area insets, which may not be appropriate for every use case.

**Comment by dlo** [comment]
@RickyRomero Got it, thanks! I see what you're seeing. What does your constraint look like?

**Comment by RickyRomero** [comment]
@dlo Here's a [simplified test case](https://github.com/lionheart/KeyboardAdjuster/files/1519803/KeyboardAdjusterTestCase.zip) which reproduces the gap behavior above (except for the swipe to dismiss; use the Done button to close the keyboard).

This test case is structured in a similar way to my own view controller, so the constraints should translate appropriately.

**Comment by dlo** [comment]
Found the issue, I just made the fix in the attached project. Basically, just add an IBOutlet for the constraint and set the constraint to the bottom of the UITextField.
[KeyboardAdjusterTestCase 2.zip](https://github.com/lionheart/KeyboardAdjuster/files/1519910/KeyboardAdjusterTestCase.2.zip)

**Comment by dlo** [comment]
<img width="263" alt="screen shot 2017-11-30 at 6 01 52 pm" src="https://user-images.githubusercontent.com/38447/33461555-8fdc3da8-d5f8-11e7-8b62-d9242c8488b8.png">
<img width="309" alt="screen shot 2017-11-30 at 6 01 46 pm" src="https://user-images.githubusercontent.com/38447/33461556-8fef4902-d5f8-11e7-92a9-d5c85f5f5a14.png">
<img width="570" alt="screen shot 2017-11-30 at 6 01 33 pm" src="https://user-images.githubusercontent.com/38447/33461557-90020d6c-d5f8-11e7-8713-6a1ea519c587.png">

**Comment by RickyRomero** [comment]
Hey Dan,

Are you testing in a newer beta than 11.2 beta 3? That appears to be the one included in the current Xcode beta simulator, and I'm still seeing the gap there with the sample you uploaded... Am I missing something?

![simulator screen shot - iphone x - 2017-11-30 at 18 26 50](https://user-images.githubusercontent.com/233891/33465309-0c56b39c-d5fc-11e7-9238-cdfc217bbdf6.png)

**Comment by dlo** [comment]
Not yet…I tested in Xcode 9.1. I'll test in the latest Xcode beta tomorrow.

**Comment by RickyRomero** [comment]
Interesting. I tested in 9.1 first and saw the same thing there, which led me to think it was something fixed in the 11.2 beta.

Thanks for all your help so far figuring this out. 🍻
